File: android/app/src/main/java/com/blaineam/haven/core/MediaOptimizationTarget.kt
package com.blaineam.haven.core

import java.io.File

/**
 * What "already optimized" looks like on Android, and how to tell by LOOKING at a file.
 *
 * The Android counterpart of Apple `MediaOptimizationTarget` (apple/HavenApp/). It is the predicate
 * [MediaReoptimizer] uses to decide which of my ALREADY-SHARED blobs are worth re-encoding: the
 * compression rewrite in [MediaTargets] only ever applied to the next thing posted, so everything
 * already out there — on my device AND on every member's — stayed exactly as big as it was.
 *
 * ## Two layers, on purpose
 *
 * The DECISION is pure: [judgeVideo] / [judgeImage] take facts (bytes, pixels, codec, duration) and
 * return a verdict. The GATHERING is not: it decrypts and reads headers ([probeVideoFile],
 * [probeImageBytes]). They are split so the decision — the part that governs how much CPU the whole
 * feature burns and whether it can loop forever — is testable on the JVM without a device, an
 * emulator, or a real video. See `MediaOptimizationTargetTest`.
 *
 * ## Idempotence is the safety property
 *
 * A shape-only gate is only sound if the encoder's OWN output re-probes as AT target. Otherwise the
 * button offers the same clip on every scan and re-encodes it forever. That is why the ceilings below
 * carry headroom over [MediaTargets]' nominal rates rather than sitting exactly on them:
 * `AVVideoAverageBitRateKey`'s Android equivalent, `MediaFormat.KEY_BIT_RATE`, is a target the encoder
 * averages towards, not a ceiling, and the container adds its own overhead on top.
 *
 * The second half of the loop guard lives in [MediaReoptimizer]: anything that fails, or comes back
 * no smaller, is added to a persisted skip set and never offered again. That is what stops a clip
 * whose re-encode legitimately can't win (already lean, but HEVC, so it reads "above target"
 * forever) from being retried on every tap.
 *
 * ## Divergences from Apple, stated plainly
 *
 * - **Audio is never a candidate.** See [judgeAudio]. Apple re-encodes standalone audio through
 *   `addAudio`; Android has no audio-file IMPORT path at all, so every `aud_` ref this device
 *   authored came from the in-app recorder, which already writes mono AAC at
 *   [MediaTargets.VOICE_NOTE_BITRATE] — provably at target. Building an AAC transcoder to re-encode
 *   files that cannot exist would be inventing work.
 * - **The video bitrate ceiling still budgets [MediaTargets.VIDEO_AUDIO_BITRATE]** even though the
 *   Android transcode COPIES the source audio track through rather than re-encoding it (see the
 *   divergence note in [MediaTargets]). Passthrough camera AAC is ordinarily 128–256 kbps, which sits
 *   comfortably inside the 1.5 Mbps of headroom the 1.30 factor leaves above the 4.5 Mbps video
 *   target. Keeping the arithmetic identical to Apple's matters more than shaving it.
 */
object MediaOptimizationTarget {

    // ---- Tolerances ---------------------------------------------------------------------------

    /**
     * Overall bits/second above which a video is worth rewriting. 30% of headroom over the nominal
     * video+audio target, for the reason in the class doc: our own output must land UNDER this or
     * the scan never converges. Measured on Apple against real encoder output: 4.5–5.0 Mbps out
     * against this 6.0 Mbps trigger.
     */
    val videoBitrateCeiling: Int =
        ((MediaTargets.VIDEO_BITRATE + MediaTargets.VIDEO_AUDIO_BITRATE) * 1.30).toInt()

    /** JPEG q62 at 1600px measures well under this. A denser still came in at a higher quality. */
    const val IMAGE_BYTES_PER_PIXEL_CEILING: Double = 0.40

    /** 2% slack purely for the even-number rounding both encoders do on output dimensions. */
    const val DIMENSION_SLACK: Double = 1.02

    /**
     * Below this there is nothing worth winning, and a re-share costs every member in the circle an
     * edit event plus a re-download. Small files are left exactly as they are.
     */
    const val MINIMUM_INTERESTING_BYTES: Long = 200_000

    /**
     * A re-encode that doesn't clearly win is worse than no re-encode: the whole circle re-downloads
     * for nothing. New bytes must be at least this much smaller to be adopted.
     */
    const val REQUIRED_SHRINK_FACTOR: Double = 0.90

    /**
     * The moment the bitrate-controlled encoder landed in Android's posting path
     * (2026-07-20 08:00 America/Los_Angeles, the same instant Apple pins). Media shared before this
     * CANNOT have come from it.
     *
     * REPORTED, never enforced — see [MediaReoptimizer.scan]. Age would wrongly exclude media shared
     * after the cutoff with auto-optimize off (half the population this button exists for) and
     * wrongly include media that is already at target. Asking the file beats asking the calendar.
     * It is kept because it explains a row to the user ("shared before Haven learned to compress").
     */
    const val LEGACY_CUTOFF_MS: Long = 1_784_559_600_000L

    fun isLegacyByAge(createdAtMs: Long): Boolean = createdAtMs < LEGACY_CUTOFF_MS

    // ---- The verdict --------------------------------------------------------------------------

    /** What a probe found. [aboveTargetReason] is null when the file already reads as at-target. */
    data class Shape(
        val bytes: Long,
        /** Longest edge in pixels; 0 for audio. */
        val maxDimension: Int,
        /** Codec/container tell — a track MIME ("video/avc") or an image MIME ("image/jpeg"). */
        val codec: String,
        /** Overall bits/second; 0 for stills. */
        val bitrate: Int,
        val seconds: Double,
        val aboveTargetReason: String?,
    ) {
        val aboveTarget: Boolean get() = aboveTargetReason != null
    }

    /**
     * CODEC is the strongest tell there is: the optimized path emits H.264 and nothing else (see
     * `transcodeVideo` — `video/avc` is hard-coded, because iOS-shot HEVC cannot be relied on to
     * decode everywhere). So a `video/hevc` track is positive proof the file came from the
     * passthrough strip-remux or the raw-copy fallback, i.e. auto-optimize was off or the transcode
     * failed. Dimensions and bitrate then catch H.264 files that came from the OLD Android path,
     * whose "adaptive" 4-bits-per-pixel formula pinned every 1080p clip to its own 8 Mbps ceiling.
     *
     * Bitrate is computed from FILE BYTES ÷ DURATION rather than any declared rate, because that is
     * the number that actually costs the circle storage and transfer, container overhead and all.
     */
    fun judgeVideo(bytes: Long, width: Int, height: Int, trackMime: String, seconds: Double): Shape? {
        if (bytes <= 0 || width <= 0 || height <= 0 || !seconds.isFinite() || seconds <= 0.0) return null
        val maxDim = maxOf(width, height)
        val bitrate = (bytes * 8.0 / seconds).toInt()
        val reason = when {
            bytes < MINIMUM_INTERESTING_BYTES -> null
            !trackMime.equals("video/avc", ignoreCase = true) -> "$trackMime, not H.264"
            maxDim > MediaTargets.VIDEO_LONG_EDGE * DIMENSION_SLACK ->
                "${maxDim}px, target ${MediaTargets.VIDEO_LONG_EDGE}px"
            bitrate > videoBitrateCeiling ->
                "${bitrate / 1000} kbps, target ≤ ${videoBitrateCeiling / 1000} kbps"
            else -> null
        }
        return Shape(bytes, maxDim, trackMime, bitrate, seconds, reason)
    }

    /**
     * Two tells, and both matter. DIMENSIONS catch the obvious case — a 4032px camera original that
     * never went through the downscale. BYTES-PER-PIXEL catch the subtler one: a photo that IS
     * 1600px but was written at q95 (auto-optimize off) or arrived as a PNG/WebP screenshot, where
     * the pixel count says nothing and only the density gives it away.
     */
    fun judgeImage(bytes: Long, width: Int, height: Int, mime: String): Shape? {
        if (bytes <= 0 || width <= 0 || height <= 0) return null
        val maxDim = maxOf(width, height)
        val bpp = bytes.toDouble() / (width.toDouble() * height.toDouble())
        val reason = when {
            bytes < MINIMUM_INTERESTING_BYTES -> null
            // The optimized path ALWAYS writes JPEG. Anything else here came in verbatim.
            !mime.equals("image/jpeg", ignoreCase = true) -> "not a JPEG ($mime)"
            maxDim > MediaTargets.STILL_LONG_EDGE * DIMENSION_SLACK ->
                "${maxDim}px, target ${MediaTargets.STILL_LONG_EDGE}px"
            bpp > IMAGE_BYTES_PER_PIXEL_CEILING ->
                String.format("%.2f bytes/pixel, target ≤ %.2f", bpp, IMAGE_BYTES_PER_PIXEL_CEILING)
            else -> null
        }
        return Shape(bytes, maxDim, mime, 0, 0.0, reason)
    }

    /**
     * Always at target. Not an oversight and not a stub — see the divergence note in the class doc:
     * on Android there is no import path that can mint a non-conforming `aud_` blob, so an audio
     * candidate cannot exist, and a "candidate" the encoder would have to refuse is worse than none
     * (it would burn a slot in the batch and a slot in the skip set on every scan).
     */
    fun judgeAudio(bytes: Long): Shape = Shape(bytes, 0, "audio", 0, 0.0, null)

    /**
     * Whether a fresh encode is worth adopting. Applied to the REAL output size against the REAL
     * source size, never to a declared rate: [MediaTargets.VIDEO_BITRATE] is a target, not a ceiling
     * on the source, so an already-lean clip is re-encoded UP and grows. Android hit exactly this on
     * a 720p fixture (0.48 MB in, 0.59 MB out) — caught by measurement, not by review.
     */
    fun keepsNewEncode(sourceBytes: Long, newBytes: Long): Boolean =
        newBytes > 0 && sourceBytes > 0 && newBytes < sourceBytes * REQUIRED_SHRINK_FACTOR

    // ---- Gathering the facts (device-only) ----------------------------------------------------

    /**
     * Probe a DECRYPTED video file. Returns null when it cannot be read or judged at all — the caller
     * leaves such blobs alone rather than guessing (fail closed: an unreadable file is not a
     * re-encode candidate).
     */
    fun probeVideoFile(file: File): Shape? {
        val bytes = file.length()
        if (bytes <= 0) return null
        var extractor: android.media.MediaExtractor? = null
        return try {
            extractor = android.media.MediaExtractor().apply { setDataSource(file.absolutePath) }
            var mime: String? = null
            var w = 0
            var h = 0
            for (i in 0 until extractor.trackCount) {
                val f = extractor.getTrackFormat(i)
                val m = f.getString(android.media.MediaFormat.KEY_MIME) ?: continue
                if (!m.startsWith("video/")) continue
                mime = m
                w = f.getInteger(android.media.MediaFormat.KEY_WIDTH)
                h = f.getInteger(android.media.MediaFormat.KEY_HEIGHT)
                break
            }
            if (mime == null) return null
            val seconds = android.media.MediaMetadataRetriever().use { r ->
                r.setDataSource(file.absolutePath)
                r.extractMetadata(android.media.MediaMetadataRetriever.METADATA_KEY_DURATION)
                    ?.toLongOrNull()?.let { it / 1000.0 }
            } ?: return null
            judgeVideo(bytes, w, h, mime, seconds)
        } catch (t: Throwable) {
            android.util.Log.w("Reoptimize", "video probe failed: ${t.message}")
            null
        } finally {
            runCatching { extractor?.release() }
        }
    }

    /** Probe DECRYPTED image bytes. Header-only decode — no full bitmap is ever allocated. */
    fun probeImageBytes(bytes: ByteArray): Shape? = runCatching {
        val bounds = android.graphics.BitmapFactory.Options().apply { inJustDecodeBounds = true }
        android.graphics.BitmapFactory.decodeByteArray(bytes, 0, bytes.size, bounds)
        judgeImage(bytes.size.toLong(), bounds.outWidth, bounds.outHeight, bounds.outMimeType ?: "unknown")
    }.getOrNull()
}

File: apple/HavenApp/MediaOptimizationTarget.swift
import AVFoundation
import CoreGraphics
import Foundation
import ImageIO

/// What "already optimized" looks like, and how to tell by LOOKING at a file.
///
/// Today's encoder rewrite gave the posting path an explicit target (see `VideoEncoder`): 1080p
/// H.264 @ 4.5 Mbps, stills at 1600px JPEG q0.62, audio AAC 96k. Everything shared before that —
/// and anything shared at any time with auto-optimize OFF — went out at whatever the camera or the
/// sender's file happened to be. A real device is sitting on 53 items / 1.3 GB with single videos at
/// 320 MB, all from the old path. This is the predicate the "re-optimize" run uses to find them.
///
/// Deliberately dependency-free (AVFoundation/ImageIO/Foundation only — no MediaStore, no FeedStore,
/// no SwiftUI), for the same reason `VideoEncoder` is: it can be compiled and RUN standalone with
/// `swiftc` against real files in haven-media before it is trusted to decide what gets rewritten.
/// The last encoder that skipped that step was wired into the posting path unrun and deadlocked on
/// every clip with audio.
enum MediaOptimizationTarget {

    // MARK: - The targets
    //
    // Stills live here rather than in MediaStore so the probe and the producer cannot drift:
    // MediaStore.optimizedImageMaxDimension/Quality alias these. Video targets come straight from
    // `VideoEncoder`, which is the thing that actually emits them.

    static let imageMaxDimension: CGFloat = 1600
    static let imageQuality: CGFloat = 0.62
    static let videoMaxDimension: CGFloat = 1920

    // MARK: - Where "legacy" starts
    //
    // 2026-07-20 08:00 America/Los_Angeles: the moment the bitrate-controlled encoder landed in the
    // posting path. Anything created before this instant CANNOT have come from it.
    //
    // Built from components rather than a magic epoch number so the intent survives being read, and
    // pinned to the wall clock of the machine the change was made on — the cutoff is a fact about
    // this repository's history, not about the reader's time zone.
    static let legacyCutoff: Date = {
        var c = DateComponents()
        c.year = 2026; c.month = 7; c.day = 20; c.hour = 8; c.minute = 0
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "America/Los_Angeles") ?? .gmt
        return cal.date(from: c) ?? Date(timeIntervalSince1970: 1_784_559_600)   // == 2026-07-20 15:00Z
    }()

    /// Was this posted before the encoder rewrite?
    static func isLegacyByAge(createdAtMs: UInt64) -> Bool {
        Date(timeIntervalSince1970: Double(createdAtMs) / 1000) < legacyCutoff
    }

    // MARK: - Tolerances
    //
    // `AVVideoAverageBitRateKey` is a target the encoder averages towards, not a ceiling — a
    // high-motion clip legitimately overshoots, and the container adds its own overhead on top. If
    // the probe's threshold sat at the nominal rate, our OWN freshly-encoded output would come back
    // reading "above target" and the button would offer the same clip forever, re-encoding it on
    // every run. The headroom below is what makes a second pass a no-op, and it is measured, not
    // guessed: real encoder output lands around 4.5–5.0 Mbps overall against a 6.0 Mbps trigger.
    static let videoBitrateCeiling = Int(Double(VideoEncoder.videoBitrate + VideoEncoder.audioBitrate) * 1.30)
    /// Same idea for stills: JPEG q0.62 at 1600px measures well under 0.40 bytes per pixel.
    static let imageBytesPerPixelCeiling: Double = 0.40
    static let audioBitrateCeiling = Int(Double(VideoEncoder.standaloneAudioBitrate) * 1.60)
    /// Dimensions get 2% slack purely for even-number rounding in the encoder.
    static let dimensionSlack: CGFloat = 1.02

    /// Below this there is nothing worth winning, and a re-share costs every member in the circle an
    /// edit event plus a re-download. Small files are left exactly as they are.
    static let minimumInterestingBytes: Int64 = 200_000

    /// A re-encode that doesn't clearly win is worse than no re-encode: the whole circle re-downloads
    /// for nothing. The new bytes must be at least this much smaller to be adopted.
    static let requiredShrinkFactor: Double = 0.90

    // MARK: - The probe

    struct Shape: Sendable {
        var bytes: Int64
        /// Longest edge in pixels (0 for audio).
        var maxDimension: Int
        /// Container/codec four-char code as text ("avc1", "hvc1", "aac ") or an ImageIO UTI for stills.
        var codec: String
        /// Overall bits/second (0 for stills).
        var bitrate: Int
        var seconds: Double
        /// Nil when the file reads as already-at-target; otherwise WHY it is being rewritten.
        var aboveTargetReason: String?

        var aboveTarget: Bool { aboveTargetReason != nil }
        var summary: String {
            let mb = String(format: "%.1f", Double(bytes) / 1_048_576)
            if bitrate > 0 {
                return "\(mb) MB · \(codec) · \(maxDimension)px · \(bitrate / 1000) kbps"
            }
            return maxDimension > 0 ? "\(mb) MB · \(codec) · \(maxDimension)px" : "\(mb) MB · \(codec)"
        }
    }

    /// Inspect a stored blob. Kind comes from the FILE EXTENSION because that is what `MediaStore`
    /// guarantees on disk (`<ref>.jpg` / `.mp4` / `.m4a`) — and because deriving it here rather than
    /// taking a `MediaKind` parameter is what keeps this file free of any app dependency.
    ///
    /// Returns nil when the file cannot be read or judged at all; the caller leaves such blobs alone
    /// rather than guessing (fail closed — an unreadable file is not a re-encode candidate).
    static func probe(_ url: URL) async -> Shape? {
        guard let bytes = (try? url.resourceValues(forKeys: [.fileSizeKey]))?.fileSize,
              bytes > 0 else { return nil }
        switch url.pathExtension.lowercased() {
        case "jpg", "jpeg", "png", "heic":  return probeImage(url, bytes: Int64(bytes))
        case "mp4", "mov", "m4v":           return await probeVideo(url, bytes: Int64(bytes))
        case "m4a", "wav", "aiff", "caf", "mp3": return await probeAudio(url, bytes: Int64(bytes))
        default:                            return nil
        }
    }

    // MARK: Stills
    //
    // Two tells, and both matter. DIMENSIONS catch the obvious case — a 4032px camera original that
    // never went through the downscale. BYTES-PER-PIXEL catch the subtler one: a photo that IS
    // 1600px but was written at q0.95 (auto-optimize off) or arrived as a PNG/HEIC screenshot, where
    // the pixel count says nothing and only the density gives it away.
    private static func probeImage(_ url: URL, bytes: Int64) -> Shape? {
        guard let src = CGImageSourceCreateWithURL(url as CFURL, nil),
              let props = CGImageSourceCopyPropertiesAtIndex(src, 0, nil) as? [CFString: Any],
              let w = props[kCGImagePropertyPixelWidth] as? Int,
              let h = props[kCGImagePropertyPixelHeight] as? Int,
              w > 0, h > 0 else { return nil }
        let type = (CGImageSourceGetType(src) as String?) ?? "unknown"
        let maxDim = max(w, h)
        let bpp = Double(bytes) / Double(w * h)
        var shape = Shape(bytes: bytes, maxDimension: maxDim, codec: shortType(type),
                          bitrate: 0, seconds: 0, aboveTargetReason: nil)
        guard bytes >= minimumInterestingBytes else { return shape }
        if !type.hasSuffix("jpeg") {
            // The optimized path ALWAYS writes JPEG. Anything else here came in verbatim.
            shape.aboveTargetReason = "not a JPEG (\(shape.codec))"
        } else if CGFloat(maxDim) > imageMaxDimension * dimensionSlack {
            shape.aboveTargetReason = "\(maxDim)px, target \(Int(imageMaxDimension))px"
        } else if bpp > imageBytesPerPixelCeiling {
            shape.aboveTargetReason = String(format: "%.2f bytes/pixel, target ≤ %.2f",
                                             bpp, imageBytesPerPixelCeiling)
        }
        return shape
    }

    // MARK: Video
    //
    // CODEC is the strongest tell there is: the optimized path emits H.264 and nothing else (see
    // VideoEncoder — AVVideoCodecKey is hard-coded, because Android cannot be relied on to decode
    // HEVC). So an `hvc1`/`hev1` track is positive proof the file came from the passthrough remux or
    // the raw-copy fallback, i.e. auto-optimize was off or every export failed. Dimensions and
    // bitrate then catch H.264 files that came from the old preset export, which capped size but
    // picked its own ~8 Mbps rate.
    //
    // Bitrate is computed from FILE BYTES ÷ DURATION rather than `estimatedDataRate`, because that
    // is the number that actually costs the circle storage and transfer, container overhead and all.
    private static func probeVideo(_ url: URL, bytes: Int64) async -> Shape? {
        let asset = AVURLAsset(url: url)
        guard let track = try? await asset.loadTracks(withMediaType: .video).first,
              let natural = try? await track.load(.naturalSize),
              let xf = try? await track.load(.preferredTransform),
              let dur = try? await asset.load(.duration) else { return nil }
        let seconds = dur.seconds
        guard seconds.isFinite, seconds > 0 else { return nil }
        let disp = natural.applying(xf)
        let maxDim = Int(max(abs(disp.width), abs(disp.height)).rounded())
        var codec = "unknown"
        if let fd = (try? await track.load(.formatDescriptions))?.first {
            codec = fourCC(CMFormatDescriptionGetMediaSubType(fd))
        }
        let bitrate = Int(Double(bytes) * 8 / seconds)
        var shape = Shape(bytes: bytes, maxDimension: maxDim, codec: codec,
                          bitrate: bitrate, seconds: seconds, aboveTargetReason: nil)
        guard bytes >= minimumInterestingBytes else { return shape }
        if codec != "avc1" {
            shape.aboveTargetReason = "\(codec), not H.264"
        } else if CGFloat(maxDim) > videoMaxDimension * dimensionSlack {
            shape.aboveTargetReason = "\(maxDim)px, target \(Int(videoMaxDimension))px"
        } else if bitrate > videoBitrateCeiling {
            shape.aboveTargetReason = "\(bitrate / 1000) kbps, target ≤ \(videoBitrateCeiling / 1000) kbps"
        }
        return shape
    }

    // MARK: Audio
    //
    // `addAudio` was a straight `copyItem` until today, so anything shared IN could be WAV/AIFF/ALAC
    // — uncompressed speech at tens of megabytes. Non-AAC is the tell; a fat AAC is caught on rate.
    private static func probeAudio(_ url: URL, bytes: Int64) async -> Shape? {
        let asset = AVURLAsset(url: url)
        guard let track = (try? await asset.loadTracks(withMediaType: .audio))?.first,
              let dur = try? await asset.load(.duration) else { return nil }
        let seconds = dur.seconds
        guard seconds.isFinite, seconds > 0 else { return nil }
        var codec = "unknown"
        if let fd = (try? await track.load(.formatDescriptions))?.first {
            codec = fourCC(CMFormatDescriptionGetMediaSubType(fd))
        }
        let bitrate = Int(Double(bytes) * 8 / seconds)
        var shape = Shape(bytes: bytes, maxDimension: 0, codec: codec,
                          bitrate: bitrate, seconds: seconds, aboveTargetReason: nil)
        guard bytes >= minimumInterestingBytes else { return shape }
        // 'aac ' is the plain AAC-LC subtype; the AAC family shares the "aac" stem.
        if !codec.hasPrefix("aac") {
            shape.aboveTargetReason = "\(codec), not AAC"
        } else if bitrate > audioBitrateCeiling {
            shape.aboveTargetReason = "\(bitrate / 1000) kbps, target ≤ \(audioBitrateCeiling / 1000) kbps"
        }
        return shape
    }

    private static func fourCC(_ code: FourCharCode) -> String {
        let bytes = [UInt8((code >> 24) & 0xFF), UInt8((code >> 16) & 0xFF),
                     UInt8((code >> 8) & 0xFF), UInt8(code & 0xFF)]
        return String(bytes: bytes, encoding: .ascii) ?? "????"
    }
    /// "public.jpeg" → "jpeg", so the reason strings stay readable.
    private static func shortType(_ uti: String) -> String {
        uti.split(separator: ".").last.map(String.init) ?? uti
    }
}
